drivers: i2c: I2C shim for nrfx TWI and TWIM drivers

Mierunski opened the pull request:
I prepared shim implementation for TWI and TWIM nrfx drivers. Changed board definitions to point to the new shim and removed old implementation.
Tested with HTS221 example on nrf52_pca20020 board.

Mierunski committed:
drivers: i2c: Add shims for nrfx TWI and TWIM drivers

Changes add a translation layer to make nrfx TWI and TWIM drivers
work with Zephyr API.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski committed:
boards: nrf: Use i2c drivers from nrfx

Changed board definitions to use TWI and TWIM drivers provided by nrfx.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski committed:
drivers: i2c: Removed redundant i2c_nrf5 shim

After switching to TWI and TWIM drivers, old i2c_nrf5 shim
is no longer needed.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

galak commented on drivers/i2c/Kconfig.nrfx:
```diff
@@ -0,0 +1,95 @@
+
+endchoice
+
+config I2C_0_NRF_SDA_PIN
```
Can we move to getting these gpio's from device tree instead.

Mierunski replied:
Moved pin settings to device tree, but also added an option to override those settings through kconfig - some DKs do not have specified pins and can be changed according to users needs.

galak reviewed (changes requested):
Since we are reworking the driver, lets use DTS for GPIOs related to SDA/SCL pins.

anangl commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,156 @@
+#define I2C_NRFX_TWI_DEVICE(idx)					\
+	static int twi_##idx##_init(struct device *dev)			\
+	{								\
+		IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_TWI##idx),		\
```
Since we've got DTS entries for I2C, use `CONFIG_I2C_##idx##_IRQ` to get the IRQ number.

Mierunski replied:
Done

Mierunski committed:
boards: nrf: Moved SDA and SCL pin configuration to DTS for nRF boards

SDA and SCL pins can now be configured through DTS.
Pins on development kits have been assigned according to arduino
headers.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

Mierunski commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,156 @@
+		nrfx_err_t res = nrfx_twi_xfer(&get_dev_config(dev)->twi,
+					       &cur_xfer,
+					       (msgs[i].flags & I2C_MSG_STOP) ?
+					       NRFX_TWI_FLAG_TX_NO_STOP : 0);
```
Fixed

Mierunski commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,156 @@
+	return dev->driver_data;
+}
+static inline const struct i2c_nrfx_twi_config
+*get_dev_config(struct device *dev)
```
Fixed

Mierunski commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,156 @@
+
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		inst->p_twi->FREQUENCY = NRF_TWI_FREQ_100K;
```
Done

anangl commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,165 @@
+
+		k_sem_take(&(get_dev_data(dev)->sync), K_FOREVER);
+		if (get_dev_data(dev)->res != NRFX_SUCCESS) {
+			return -EIO;
```
What about adding something like
`SYS_LOG_ERR("Error %d occurred for message %d", get_dev_data(dev)->res, i);`
one line above?

`res` could be copied to a local variable to avoid re-read (this field is volatile).

Mierunski replied:
Done

anangl commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,165 @@
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&(get_dev_data(dev)->sync));
```
`&dev_data->sync`

Mierunski replied:
Fixed

anangl commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -0,0 +1,165 @@
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&(get_dev_data(dev)->sync));
```
`&dev_data->sync`

Mierunski replied:
Fixed

anangl commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -0,0 +1,165 @@
+	return 0;
+}
+
+#define I2C_NRFX_TWI_DEVICE(idx)					\
```
`I2C_NRFX_TWIM_DEVICE`

Mierunski replied:
Fixed

anangl commented on ext/hal/nordic/Kconfig:
```diff
@@ -21,3 +21,9 @@ config NRFX_SPIS
+config NRFX_TWI
+	bool
+
+config NRFX_TWIM
```
Place these entries before `NRFX_WDT` to keep the alphabetical order.

anangl commented on drivers/i2c/Kconfig.nrfx:
```diff
@@ -0,0 +1,115 @@
+	depends on SOC_SERIES_NRF52X
+	select NRFX_TWIM
+	help
+	  Enable nRF TWI Master with EasyDMA on port 0.
```
It would be good to mention here that this peripheral accepts only transfers from RAM, and when a buffer placed in flash is provided, the transfer will fail.

Mierunski replied:
Done

anangl commented on drivers/i2c/Kconfig.nrfx:
```diff
@@ -0,0 +1,115 @@
+
+if I2C_NRFX
+
+config I2C_NRFX_PINS_OVERRIDE_DTS
```
Wouldn't `I2C_NRF_OVERRIDE_DTS_PINS` be more adequate (NRF, because this entry, like pins settings, is not nrfx related but generic for nRF chips)?

Mierunski replied:
Changed

anangl reviewed (changes requested):
A few more issues, mostly minor ones.

kl-cruz commented:
I think this PR should trace changes in https://github.com/zephyrproject-rtos/zephyr/pull/7915/files#r192312703

Mierunski commented on boards/arm/bbc_microbit/Kconfig.defconfig:
```diff
@@ -28,22 +28,31 @@ endif # UART_NRF5
-config I2C_NRF5_0_GPIO_SCL_PIN
-	default 0
+choice I2C_0_NRF_TYPE
+	default I2C_0_NRF_TWIM
```
Changed to TWI

Mierunski commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2017 - 2018, Nordic Semiconductor ASA
```
Fixed

tbursztyka commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,160 @@
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&dev_data->sync);
```
add an empty line before

Mierunski replied:
done

tbursztyka commented on drivers/i2c/i2c_nrfx_twi.c:
```diff
@@ -0,0 +1,160 @@
+			    dev->config->name);
+		return -EBUSY;
+	}
+	nrfx_twi_enable(&get_dev_config(dev)->twi);
```
same

tbursztyka commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -0,0 +1,160 @@
+			return -EIO;
+		}
+	}
+	return 0;
```
same

tbursztyka commented on drivers/i2c/i2c_nrfx_twim.c:
```diff
@@ -0,0 +1,160 @@
+		dev_data->res = NRFX_ERROR_INTERNAL;
+		break;
+	}
+	k_sem_give(&dev_data->sync);
```
etc...

anangl commented on boards/arm/nrf52_pca20020/board.c:
```diff
@@ -12,6 +12,8 @@
 #define CCS_VDD_PWR_CTRL_GPIO_PIN 10
 
+#ifdef CONFIG_GPIO
+
```
This change does not seem to be related to TWI/TWIM shims or I2C DTS entries, so it should not be part of this PR.
And is it needed at all?

anangl replied:
On second thought, I think this issue should be solved in a different way. The GPIO module should be explicitly enabled in the default configuration for pca20020 (previously, it was enabled implicitly by the old I2C driver), so that the code in `board.c` could turn on the VDD power rail. And this power rail is needed for example for the SX1509B GPIO expander to operate, and this expander is enabled by default for this board (and it depends on GPIO as well). Thus, I think the assumption of the author of this part of code was that GPIO should be always enabled (hence, CONFIG_GPIO was not checked). @tkln Could you confirm my understanding is correct?

Mierunski replied:
Done

anangl commented on drivers/i2c/Kconfig.nrfx:
```diff
@@ -0,0 +1,117 @@
+	depends on SOC_SERIES_NRF52X
+	select NRFX_TWIM
+	help
+	  Enable nRF TWI Master with EasyDMA on port 1.
```
Add the same note regarding transfers from RAM as for I2C_0_NRF_TWIM.

Mierunski replied:
Done

anangl reviewed (changes requested):
Change the title of the 4th commit to "boards: nrf: Moved SDA and SCL pin configuration to DTS".
And in 1st commit "shim" -> "shims" (but it's just a detail).

Mierunski committed:
boards: nrf: Fix nrf52_pca20020 board for benchmark test.

Currently tests/benchmark/footprint build is failing for
nrf52_pca20020 board. In minimal configuration, all modules are
disabled, including GPIO, because of that defines related to GPIO
are missing in board.c file and compilation fails.

To solve it, I selected GPIO to be always turned on for
nrf52_pca20020 board.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

carlescufi commented:
@galak can you revisit this now that the release is out of the way?

carlescufi reviewed (approved):
LGTM now

galak commented on drivers/i2c/Kconfig.nrfx:
```diff
@@ -0,0 +1,119 @@
+
+if I2C_NRFX
+
+config I2C_NRF_OVERRIDE_DTS_PINS
```
Why do we need this?

Mierunski replied:
It is mainly used for development kits, where I2C pins can be changed quite often, depending on hardware connected to the board.
We can't use configuration from DTS as default pins for Kconfig, but by adding override option we can switch to using pin configuration from DTS or to using user configuration from menuconfig.

galak replied:
> It is mainly used for development kits, where I2C pins can be changed quite often, depending on hardware connected to the board.
> We can't use configuration from DTS as default pins for Kconfig, but by adding override option we can switch to using pin configuration from DTS or to using user configuration from menuconfig.

I'm not keen on having support for this, we should be doing HW cfg via the DTS and only DTS.

carlescufi replied:
Actually since there is an active discussion regarding Kconfig and DT I'll copy a couple of people here who have been involved so we get more perspectives: @erwango @tbursztyka @SebastianBoe

erwango replied:
think having multiple ways for HW configuration (or anything in general) is not a good idea.
Some will use one way, some will use other, it'll soon become difficult to maintain.

tbursztyka replied:
Aren't sda/scd pins the same on 1 board, no matter how many slaves you wire up?

jarz-nordic replied:
> Aren't sda/scd pins the same on 1 board, no matter how many slaves you wire up?

No, in NRF SoCs you can connect SDA/SDC PINs wherevere you wish and you can modify them during runtime.
I saw that for instance ST can have 2 sets of PINs for UART in DTS. We can be more flexible thus we would like to have  possibility to overwride DTS in Kconfig. Putting all possible combinations to DTS makes limited sense in my opinion.

carlescufi commented:
@galak your request was addressed from what I can tell, can your revisit?
